File: services/event-streaming-service/src/handlers/branches.rs
//! Stream branching REST surface (Bloque E1).
//!
//! Exposes 6 endpoints under `/api/v1/streaming/streams/{stream_id}/branches`:
//!
//! | Verb   | Path                                | Purpose                        |
//! |--------|-------------------------------------|--------------------------------|
//! | GET    | `/branches`                         | list                           |
//! | POST   | `/branches`                         | create                         |
//! | GET    | `/branches/{name}`                  | fetch single branch            |
//! | DELETE | `/branches/{name}`                  | hard-delete (only when empty)  |
//! | POST   | `/branches/{name}:merge`            | merge into target branch       |
//! | POST   | `/branches/{name}:archive`          | mark archived & optionally     |
//! |        |                                     | commit cold tier               |
//!
//! Cold-branch materialisation delegates to `dataset-versioning-service`
//! over plain HTTP using `state.http_client` and the URL configured via
//! [`crate::AppState::dataset_service_url`]. Failures only surface in the
//! response message — the local row is always written first.

use auth_middleware::claims::Claims;
use axum::{
    Extension, Json,
    extract::{Path, State},
};
use chrono::Utc;
use sqlx::{Postgres, Transaction, types::Json as SqlJson};
use uuid::Uuid;

use crate::{
    AppState,
    handlers::{
        ServiceResult, bad_request, db_error, forbidden, not_found, streams::emit_audit_event,
    },
    models::{
        ListResponse,
        branch::{
            ArchiveBranchRequest, CreateBranchRequest, MergeBranchRequest, MergeBranchResponse,
            StreamBranch, StreamBranchRow,
        },
    },
    outbox as streaming_outbox,
};

const PERM_BRANCH_WRITE: &str = "streaming:write";

fn can_write(claims: &Claims) -> bool {
    claims.has_any_role(&["admin", "streaming_admin", "data_engineer"])
        || claims.has_permission_key(PERM_BRANCH_WRITE)
}

async fn ensure_stream_exists(db: &sqlx::PgPool, stream_id: Uuid) -> Result<(), sqlx::Error> {
    let exists: bool =
        sqlx::query_scalar("SELECT EXISTS(SELECT 1 FROM streaming_streams WHERE id = $1)")
            .bind(stream_id)
            .fetch_one(db)
            .await?;
    if exists {
        Ok(())
    } else {
        Err(sqlx::Error::RowNotFound)
    }
}

async fn load_branch_by_name(
    db: &sqlx::PgPool,
    stream_id: Uuid,
    name: &str,
) -> Result<StreamBranchRow, sqlx::Error> {
    sqlx::query_as::<_, StreamBranchRow>(
        "SELECT id, stream_id, name, parent_branch_id, status, head_sequence_no,
                dataset_branch_id, description, created_by, created_at, archived_at
           FROM streaming_stream_branches
          WHERE stream_id = $1 AND name = $2",
    )
    .bind(stream_id)
    .bind(name)
    .fetch_one(db)
    .await
}

async fn load_branch_by_name_tx(
    tx: &mut Transaction<'_, Postgres>,
    stream_id: Uuid,
    name: &str,
) -> Result<StreamBranchRow, sqlx::Error> {
    sqlx::query_as::<_, StreamBranchRow>(
        "SELECT id, stream_id, name, parent_branch_id, status, head_sequence_no,
                dataset_branch_id, description, created_by, created_at, archived_at
           FROM streaming_stream_branches
          WHERE stream_id = $1 AND name = $2",
    )
    .bind(stream_id)
    .bind(name)
    .fetch_one(&mut **tx)
    .await
}

pub async fn list_branches(
    State(state): State<AppState>,
    Path(stream_id): Path<Uuid>,
) -> ServiceResult<ListResponse<StreamBranch>> {
    if let Err(sqlx::Error::RowNotFound) = ensure_stream_exists(&state.db, stream_id).await {
        return Err(not_found("stream not found"));
    }
    let rows = sqlx::query_as::<_, StreamBranchRow>(
        "SELECT id, stream_id, name, parent_branch_id, status, head_sequence_no,
                dataset_branch_id, description, created_by, created_at, archived_at
           FROM streaming_stream_branches
          WHERE stream_id = $1
          ORDER BY created_at ASC",
    )
    .bind(stream_id)
    .fetch_all(&state.db)
    .await
    .map_err(|cause| db_error(&cause))?;
    Ok(Json(ListResponse {
        data: rows.into_iter().map(Into::into).collect(),
    }))
}

pub async fn create_branch(
    State(state): State<AppState>,
    Extension(claims): Extension<Claims>,
    Path(stream_id): Path<Uuid>,
    Json(payload): Json<CreateBranchRequest>,
) -> ServiceResult<StreamBranch> {
    if !can_write(&claims) {
        return Err(forbidden("caller lacks 'streaming:write' permission"));
    }
    let name = payload.name.trim();
    if name.is_empty() {
        return Err(bad_request("branch name is required"));
    }
    if !name
        .chars()
        .all(|c| c.is_ascii_alphanumeric() || c == '-' || c == '_' || c == '/')
    {
        return Err(bad_request(
            "branch name must contain only alphanumerics, '-', '_' or '/'",
        ));
    }
    if let Err(sqlx::Error::RowNotFound) = ensure_stream_exists(&state.db, stream_id).await {
        return Err(not_found("stream not found"));
    }
    // Parent branch (when provided) must exist and belong to the same stream.
    if let Some(parent_id) = payload.parent_branch_id {
        let parent_belongs: bool = sqlx::query_scalar(
            "SELECT EXISTS(SELECT 1 FROM streaming_stream_branches
                            WHERE id = $1 AND stream_id = $2)",
        )
        .bind(parent_id)
        .bind(stream_id)
        .fetch_one(&state.db)
        .await
        .map_err(|cause| db_error(&cause))?;
        if !parent_belongs {
            return Err(bad_request("parent_branch_id is not part of this stream"));
        }
    }

    let id = Uuid::now_v7();
    let mut tx = state.db.begin().await.map_err(|cause| db_error(&cause))?;
    sqlx::query(
        "INSERT INTO streaming_stream_branches (
             id, stream_id, name, parent_branch_id, status, head_sequence_no,
             dataset_branch_id, description, created_by
         ) VALUES ($1, $2, $3, $4, 'active', 0, $5, $6, $7)",
    )
    .bind(id)
    .bind(stream_id)
    .bind(name)
    .bind(payload.parent_branch_id)
    .bind(payload.dataset_branch_id.as_deref())
    .bind(payload.description.unwrap_or_default())
    .bind(&claims.email)
    .execute(&mut *tx)
    .await
    .map_err(|cause| db_error(&cause))?;

    let row = load_branch_by_name_tx(&mut tx, stream_id, name)
        .await
        .map_err(|cause| db_error(&cause))?;
    let branch: StreamBranch = row.into();
    streaming_outbox::emit(&mut tx, &streaming_outbox::branch_created(&branch))
        .await
        .map_err(|cause| {
            tracing::error!(branch_id = %branch.id, error = %cause, "failed to enqueue outbox event");
            crate::handlers::internal_error("failed to enqueue outbox event")
        })?;
    tx.commit().await.map_err(|cause| db_error(&cause))?;
    emit_audit_event(
        &claims,
        "streaming.branch.created",
        "streaming_stream_branch",
        branch.id,
        serde_json::json!({ "stream_id": stream_id, "name": branch.name }),
    );
    Ok(Json(branch))
}

pub async fn get_branch(
    State(state): State<AppState>,
    Path((stream_id, name)): Path<(Uuid, String)>,
) -> ServiceResult<StreamBranch> {
    let row = match load_branch_by_name(&state.db, stream_id, &name).await {
        Ok(row) => row,
        Err(sqlx::Error::RowNotFound) => return Err(not_found("branch not found")),
        Err(cause) => return Err(db_error(&cause)),
    };
    Ok(Json(row.into()))
}

pub async fn delete_branch(
    State(state): State<AppState>,
    Extension(claims): Extension<Claims>,
    Path((stream_id, name)): Path<(Uuid, String)>,
) -> ServiceResult<serde_json::Value> {
    if !can_write(&claims) {
        return Err(forbidden("caller lacks 'streaming:write' permission"));
    }
    if name == "main" {
        return Err(bad_request("cannot delete the 'main' branch"));
    }
    let row = match load_branch_by_name(&state.db, stream_id, &name).await {
        Ok(row) => row,
        Err(sqlx::Error::RowNotFound) => return Err(not_found("branch not found")),
        Err(cause) => return Err(db_error(&cause)),
    };
    if row.head_sequence_no > 0 && row.status != "merged" && row.status != "archived" {
        return Err(bad_request(
            "branch has uncommitted history; archive or merge it first",
        ));
    }
    let deleted_at = Utc::now();
    let mut tx = state.db.begin().await.map_err(|cause| db_error(&cause))?;
    sqlx::query("DELETE FROM streaming_stream_branches WHERE id = $1")
        .bind(row.id)
        .execute(&mut *tx)
        .await
        .map_err(|cause| db_error(&cause))?;
    let branch: StreamBranch = row.into();
    streaming_outbox::emit(
        &mut tx,
        &streaming_outbox::branch_deleted(&branch, deleted_at, branch.head_sequence_no),
    )
    .await
    .map_err(|cause| {
        tracing::error!(branch_id = %branch.id, error = %cause, "failed to enqueue outbox event");
        crate::handlers::internal_error("failed to enqueue outbox event")
    })?;
    tx.commit().await.map_err(|cause| db_error(&cause))?;
    emit_audit_event(
        &claims,
        "streaming.branch.deleted",
        "streaming_stream_branch",
        branch.id,
        serde_json::json!({ "stream_id": stream_id, "name": name }),
    );
    Ok(Json(
        serde_json::json!({ "deleted": true, "id": branch.id }),
    ))
}

pub async fn merge_branch(
    State(state): State<AppState>,
    Extension(claims): Extension<Claims>,
    Path((stream_id, name)): Path<(Uuid, String)>,
    Json(payload): Json<MergeBranchRequest>,
) -> ServiceResult<MergeBranchResponse> {
    if !can_write(&claims) {
        return Err(forbidden("caller lacks 'streaming:write' permission"));
    }
    let target_name = payload
        .target_branch
        .clone()
        .unwrap_or_else(|| "main".to_string());
    if target_name == name {
        return Err(bad_request("target branch must differ from source"));
    }
    let source = match load_branch_by_name(&state.db, stream_id, &name).await {
        Ok(row) => row,
        Err(sqlx::Error::RowNotFound) => return Err(not_found("source branch not found")),
        Err(cause) => return Err(db_error(&cause)),
    };
    let target = match load_branch_by_name(&state.db, stream_id, &target_name).await {
        Ok(row) => row,
        Err(sqlx::Error::RowNotFound) => return Err(not_found("target branch not found")),
        Err(cause) => return Err(db_error(&cause)),
    };
    let merged_seq = source.head_sequence_no.max(target.head_sequence_no);

    let merged_at = Utc::now();
    let mut tx = state.db.begin().await.map_err(|cause| db_error(&cause))?;
    sqlx::query(
        "UPDATE streaming_stream_branches
            SET head_sequence_no = $2
          WHERE id = $1",
    )
    .bind(target.id)
    .bind(merged_seq)
    .execute(&mut *tx)
    .await
    .map_err(|cause| db_error(&cause))?;

    sqlx::query(
        "UPDATE streaming_stream_branches
            SET status = 'merged'
          WHERE id = $1",
    )
    .bind(source.id)
    .execute(&mut *tx)
    .await
    .map_err(|cause| db_error(&cause))?;
    let source_branch: StreamBranch = source.into();
    streaming_outbox::emit(
        &mut tx,
        &streaming_outbox::branch_merged(&source_branch, target.id, merged_seq, merged_at),
    )
    .await
    .map_err(|cause| {
        tracing::error!(branch_id = %source_branch.id, error = %cause, "failed to enqueue outbox event");
        crate::handlers::internal_error("failed to enqueue outbox event")
    })?;
    tx.commit().await.map_err(|cause| db_error(&cause))?;

    emit_audit_event(
        &claims,
        "streaming.branch.merged",
        "streaming_stream_branch",
        source_branch.id,
        serde_json::json!({
            "stream_id": stream_id,
            "source": name,
            "target": target_name,
            "merged_sequence_no": merged_seq,
        }),
    );

    Ok(Json(MergeBranchResponse {
        source_branch_id: source_branch.id,
        target_branch_id: target.id,
        merged_sequence_no: merged_seq,
        message: format!("merged '{}' into '{}'", name, target_name),
    }))
}

pub async fn archive_branch(
    State(state): State<AppState>,
    Extension(claims): Extension<Claims>,
    Path((stream_id, name)): Path<(Uuid, String)>,
    Json(payload): Json<ArchiveBranchRequest>,
) -> ServiceResult<StreamBranch> {
    if !can_write(&claims) {
        return Err(forbidden("caller lacks 'streaming:write' permission"));
    }
    let row = match load_branch_by_name(&state.db, stream_id, &name).await {
        Ok(row) => row,
        Err(sqlx::Error::RowNotFound) => return Err(not_found("branch not found")),
        Err(cause) => return Err(db_error(&cause)),
    };

    let mut tx = state.db.begin().await.map_err(|cause| db_error(&cause))?;
    sqlx::query(
        "UPDATE streaming_stream_branches
            SET status = 'archived',
                archived_at = now()
          WHERE id = $1",
    )
    .bind(row.id)
    .execute(&mut *tx)
    .await
    .map_err(|cause| db_error(&cause))?;
    let updated = load_branch_by_name_tx(&mut tx, stream_id, &name)
        .await
        .map_err(|cause| db_error(&cause))?;
    let updated_branch: StreamBranch = updated.into();
    streaming_outbox::emit(&mut tx, &streaming_outbox::branch_archived(&updated_branch))
        .await
        .map_err(|cause| {
            tracing::error!(branch_id = %updated_branch.id, error = %cause, "failed to enqueue outbox event");
            crate::handlers::internal_error("failed to enqueue outbox event")
        })?;
    tx.commit().await.map_err(|cause| db_error(&cause))?;

    // Best-effort cold-tier commit. We POST the branch metadata to
    // dataset-versioning-service so it can snapshot the cold copy. We
    // never fail the request on HTTP errors \u2014 the local row is already
    // archived; the operator can retry via the dataset-versioning UI.
    if payload.commit_cold {
        if let Some(dataset_branch) = updated_branch.dataset_branch_id.as_deref() {
            let url = format!(
                "{}/api/v1/datasets/{}/branches/{}:commit",
                state.dataset_service_url.trim_end_matches('/'),
                updated_branch.stream_id,
                dataset_branch,
            );
            let body = serde_json::json!({
                "stream_id": stream_id,
                "branch_name": name,
                "head_sequence_no": updated_branch.head_sequence_no,
                "marking": null,
                "archived_at": Utc::now().to_rfc3339(),
            });
            // Suppress lint: best-effort.
            let _ = SqlJson(()); // keep import live in case of future tweaks.
            match state.http_client.post(&url).json(&body).send().await {
                Ok(resp) if resp.status().is_success() => {
                    tracing::info!(branch = %updated_branch.id, "cold tier commit accepted");
                    // Bloque F1: cold archive counter (rows archived).
                    state
                        .metrics
                        .record_stream_rows_archived(&name, updated_branch.head_sequence_no as u64);
                }
                Ok(resp) => {
                    tracing::warn!(
                        branch = %updated_branch.id,
                        status = %resp.status(),
                        "cold tier commit rejected"
                    );
                }
                Err(err) => {
                    tracing::warn!(branch = %updated_branch.id, error = %err, "cold tier commit failed");
                }
            }
        } else {
            tracing::info!(branch = %updated_branch.id, "no dataset_branch_id; skipping cold commit");
        }
    }

    emit_audit_event(
        &claims,
        "streaming.branch.archived",
        "streaming_stream_branch",
        row.id,
        serde_json::json!({
            "stream_id": stream_id,
            "name": name,
            "commit_cold": payload.commit_cold,
        }),
    );
    Ok(Json(updated_branch))
}

File: services/ingestion-replication-service/internal/handlers/branches.go
package handlers

// IRF-8 — Stream branches REST surface (Bloque E1). Mirrors
// services/ingestion-replication-service/src/event_streaming/handlers/branches.rs.
//
//   GET    /api/v1/streaming/streams/{id}/branches
//   POST   /api/v1/streaming/streams/{id}/branches
//   GET    /api/v1/streaming/streams/{id}/branches/{name}
//   DELETE /api/v1/streaming/streams/{id}/branches/{name}
//   POST   /api/v1/streaming/streams/{id}/branches/{name}:merge
//   POST   /api/v1/streaming/streams/{id}/branches/{name}:archive
//
// Cold-branch materialisation delegates to dataset-versioning-service
// over HTTP using the injected ColdTierBridge — failures only surface
// in the response message (the local row is always written first).

import (
	"context"
	"encoding/json"
	"errors"
	"log/slog"
	"net/http"
	"strings"
	"time"

	"github.com/go-chi/chi/v5"
	"github.com/google/uuid"

	authmw "github.com/openfoundry/openfoundry-go/libs/auth-middleware"
	"github.com/openfoundry/openfoundry-go/services/ingestion-replication-service/internal/domain/streambranch"
)

// BranchesHandler bundles the six stream-branch endpoints.
type BranchesHandler struct {
	Store      streambranch.Store
	Cold       streambranch.ColdTierBridge
	MetricSink BranchMetricSink
}

// BranchMetricSink mirrors the Rust `metrics.record_stream_rows_archived`
// hook so the cold archive counter stays connected when the operator
// wires the production observability stack. nil disables emission.
type BranchMetricSink interface {
	RecordStreamRowsArchived(branch string, rows uint64)
}

// MainBranchName is the protected branch name. Mirrors Rust's literal
// "main" check inside delete_branch.
const MainBranchName = "main"

// ListBranches is GET /streams/{id}/branches.
func (h *BranchesHandler) ListBranches(w http.ResponseWriter, r *http.Request) {
	if _, ok := requireClaims(w, r); !ok {
		return
	}
	streamID, ok := parseStreamID(w, r)
	if !ok {
		return
	}
	exists, err := h.Store.StreamExists(r.Context(), streamID)
	if err != nil {
		writeJSONErr(w, http.StatusInternalServerError, "database operation failed")
		return
	}
	if !exists {
		writeJSONErr(w, http.StatusNotFound, "stream not found")
		return
	}
	items, err := h.Store.ListBranches(r.Context(), streamID)
	if err != nil {
		writeJSONErr(w, http.StatusInternalServerError, "database operation failed")
		return
	}
	if items == nil {
		items = []streambranch.StreamBranch{}
	}
	writeJSON(w, http.StatusOK, map[string]any{"data": items})
}

// CreateBranch is POST /streams/{id}/branches.
func (h *BranchesHandler) CreateBranch(w http.ResponseWriter, r *http.Request) {
	claims, ok := requireClaims(w, r)
	if !ok {
		return
	}
	if !canWriteStreams(claims) {
		writeJSONErr(w, http.StatusForbidden, "caller lacks 'streaming:write' permission")
		return
	}
	streamID, ok := parseStreamID(w, r)
	if !ok {
		return
	}
	var body streambranch.CreateBranchRequest
	if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
		writeJSONErr(w, http.StatusBadRequest, "invalid body")
		return
	}
	name := strings.TrimSpace(body.Name)
	if name == "" {
		writeJSONErr(w, http.StatusBadRequest, "branch name is required")
		return
	}
	if !validBranchName(name) {
		writeJSONErr(w, http.StatusBadRequest,
			"branch name must contain only alphanumerics, '-', '_' or '/'")
		return
	}
	exists, err := h.Store.StreamExists(r.Context(), streamID)
	if err != nil {
		writeJSONErr(w, http.StatusInternalServerError, "database operation failed")
		return
	}
	if !exists {
		writeJSONErr(w, http.StatusNotFound, "stream not found")
		return
	}
	if body.ParentBranchID != nil {
		ok, err := h.Store.ParentBranchBelongsTo(r.Context(), *body.ParentBranchID, streamID)
		if err != nil {
			writeJSONErr(w, http.StatusInternalServerError, "database operation failed")
			return
		}
		if !ok {
			writeJSONErr(w, http.StatusBadRequest, "parent_branch_id is not part of this stream")
			return
		}
	}
	description := ""
	if body.Description != nil {
		description = *body.Description
	}
	branch, err := h.Store.CreateBranch(
		r.Context(), streamID, name, branchActor(claims),
		body.ParentBranchID, body.DatasetBranchID, description,
	)
	if err != nil {
		slog.Error("create branch", slog.String("error", err.Error()))
		writeJSONErr(w, http.StatusInternalServerError, "database operation failed")
		return
	}
	emitBranchAudit(claims, "streaming.branch.created", branch.ID,
		slog.String("stream_id", streamID.String()),
		slog.String("name", branch.Name))
	writeJSON(w, http.StatusOK, branch)
}

// GetBranch is GET /streams/{id}/branches/{name}.
func (h *BranchesHandler) GetBranch(w http.ResponseWriter, r *http.Request) {
	if _, ok := requireClaims(w, r); !ok {
		return
	}
	streamID, name, ok := parseStreamAndName(w, r)
	if !ok {
		return
	}
	branch, err := h.Store.GetBranchByName(r.Context(), streamID, name)
	if err != nil {
		writeJSONErr(w, http.StatusInternalServerError, "database operation failed")
		return
	}
	if branch == nil {
		writeJSONErr(w, http.StatusNotFound, "branch not found")
		return
	}
	writeJSON(w, http.StatusOK, branch)
}

// DeleteBranch is DELETE /streams/{id}/branches/{name}. Hard-deletes
// the row when (a) the name is not "main", and (b) the branch is empty
// (head_sequence_no == 0) or already terminated (status in
// {merged, archived}).
func (h *BranchesHandler) DeleteBranch(w http.ResponseWriter, r *http.Request) {
	claims, ok := requireClaims(w, r)
	if !ok {
		return
	}
	if !canWriteStreams(claims) {
		writeJSONErr(w, http.StatusForbidden, "caller lacks 'streaming:write' permission")
		return
	}
	streamID, name, ok := parseStreamAndName(w, r)
	if !ok {
		return
	}
	if name == MainBranchName {
		writeJSONErr(w, http.StatusBadRequest, "cannot delete the 'main' branch")
		return
	}
	branch, err := h.Store.GetBranchByName(r.Context(), streamID, name)
	if err != nil {
		writeJSONErr(w, http.StatusInternalServerError, "database operation failed")
		return
	}
	if branch == nil {
		writeJSONErr(w, http.StatusNotFound, "branch not found")
		return
	}
	if branch.HeadSequenceNo > 0 && branch.Status != "merged" && branch.Status != "archived" {
		writeJSONErr(w, http.StatusBadRequest,
			"branch has uncommitted history; archive or merge it first")
		return
	}
	if err := h.Store.DeleteBranch(r.Context(), branch.ID); err != nil {
		slog.Error("delete branch", slog.String("error", err.Error()))
		writeJSONErr(w, http.StatusInternalServerError, "database operation failed")
		return
	}
	emitBranchAudit(claims, "streaming.branch.deleted", branch.ID,
		slog.String("stream_id", streamID.String()),
		slog.String("name", name))
	writeJSON(w, http.StatusOK, map[string]any{
		"deleted": true,
		"id":      branch.ID,
	})
}

// MergeBranch is POST /streams/{id}/branches/{name}:merge.
func (h *BranchesHandler) MergeBranch(w http.ResponseWriter, r *http.Request) {
	claims, ok := requireClaims(w, r)
	if !ok {
		return
	}
	if !canWriteStreams(claims) {
		writeJSONErr(w, http.StatusForbidden, "caller lacks 'streaming:write' permission")
		return
	}
	streamID, name, ok := parseStreamAndName(w, r)
	if !ok {
		return
	}
	var body streambranch.MergeBranchRequest
	if r.ContentLength != 0 {
		if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
			writeJSONErr(w, http.StatusBadRequest, "invalid body")
			return
		}
	}
	target := MainBranchName
	if body.TargetBranch != nil && *body.TargetBranch != "" {
		target = *body.TargetBranch
	}
	if target == name {
		writeJSONErr(w, http.StatusBadRequest, "target branch must differ from source")
		return
	}
	source, err := h.Store.GetBranchByName(r.Context(), streamID, name)
	if err != nil {
		writeJSONErr(w, http.StatusInternalServerError, "database operation failed")
		return
	}
	if source == nil {
		writeJSONErr(w, http.StatusNotFound, "source branch not found")
		return
	}
	targetBranch, err := h.Store.GetBranchByName(r.Context(), streamID, target)
	if err != nil {
		writeJSONErr(w, http.StatusInternalServerError, "database operation failed")
		return
	}
	if targetBranch == nil {
		writeJSONErr(w, http.StatusNotFound, "target branch not found")
		return
	}
	merged := source.HeadSequenceNo
	if targetBranch.HeadSequenceNo > merged {
		merged = targetBranch.HeadSequenceNo
	}
	if err := h.Store.MergeBranches(r.Context(), source.ID, targetBranch.ID, merged); err != nil {
		slog.Error("merge branches", slog.String("error", err.Error()))
		writeJSONErr(w, http.StatusInternalServerError, "database operation failed")
		return
	}
	emitBranchAudit(claims, "streaming.branch.merged", source.ID,
		slog.String("stream_id", streamID.String()),
		slog.String("source", name),
		slog.String("target", target),
		slog.Int64("merged_sequence_no", merged))
	writeJSON(w, http.StatusOK, streambranch.MergeBranchResponse{
		SourceBranchID:   source.ID,
		TargetBranchID:   targetBranch.ID,
		MergedSequenceNo: merged,
		Message:          "merged '" + name + "' into '" + target + "'",
	})
}

// ArchiveBranch is POST /streams/{id}/branches/{name}:archive.
func (h *BranchesHandler) ArchiveBranch(w http.ResponseWriter, r *http.Request) {
	claims, ok := requireClaims(w, r)
	if !ok {
		return
	}
	if !canWriteStreams(claims) {
		writeJSONErr(w, http.StatusForbidden, "caller lacks 'streaming:write' permission")
		return
	}
	streamID, name, ok := parseStreamAndName(w, r)
	if !ok {
		return
	}
	var body streambranch.ArchiveBranchRequest
	if r.ContentLength != 0 {
		if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
			writeJSONErr(w, http.StatusBadRequest, "invalid body")
			return
		}
	}
	current, err := h.Store.GetBranchByName(r.Context(), streamID, name)
	if err != nil {
		writeJSONErr(w, http.StatusInternalServerError, "database operation failed")
		return
	}
	if current == nil {
		writeJSONErr(w, http.StatusNotFound, "branch not found")
		return
	}
	updated, err := h.Store.ArchiveBranch(r.Context(), streamID, current.ID, name)
	if err != nil {
		slog.Error("archive branch", slog.String("error", err.Error()))
		writeJSONErr(w, http.StatusInternalServerError, "database operation failed")
		return
	}
	if updated == nil {
		writeJSONErr(w, http.StatusNotFound, "branch not found")
		return
	}

	// Best-effort cold-tier commit. Failures never surface — the local
	// row is already archived; operators can retry via the dataset-
	// versioning UI. Mirrors the Rust handler's warn-and-continue path.
	if body.CommitCold {
		if updated.DatasetBranchID != nil && *updated.DatasetBranchID != "" {
			h.commitColdBestEffort(r.Context(), updated)
		} else {
			slog.Info("cold tier commit skipped: no dataset_branch_id",
				slog.String("branch_id", updated.ID.String()))
		}
	}

	emitBranchAudit(claims, "streaming.branch.archived", current.ID,
		slog.String("stream_id", streamID.String()),
		slog.String("name", name),
		slog.Bool("commit_cold", body.CommitCold))
	writeJSON(w, http.StatusOK, updated)
}

func (h *BranchesHandler) commitColdBestEffort(ctx context.Context, branch *streambranch.StreamBranch) {
	if h.Cold == nil {
		slog.Info("cold tier bridge not configured; skipping cold commit",
			slog.String("branch_id", branch.ID.String()))
		return
	}
	accepted, err := h.Cold.CommitCold(ctx, branch, time.Now().UTC())
	if err != nil {
		slog.Warn("cold tier commit failed",
			slog.String("branch_id", branch.ID.String()),
			slog.String("error", err.Error()))
		return
	}
	if !accepted {
		slog.Warn("cold tier commit rejected",
			slog.String("branch_id", branch.ID.String()))
		return
	}
	slog.Info("cold tier commit accepted",
		slog.String("branch_id", branch.ID.String()))
	if h.MetricSink != nil && branch.HeadSequenceNo > 0 {
		h.MetricSink.RecordStreamRowsArchived(branch.Name, uint64(branch.HeadSequenceNo))
	}
}

func parseStreamAndName(w http.ResponseWriter, r *http.Request) (uuid.UUID, string, bool) {
	streamID, ok := parseStreamID(w, r)
	if !ok {
		return uuid.Nil, "", false
	}
	name := chi.URLParam(r, "name")
	if name == "" {
		writeJSONErr(w, http.StatusBadRequest, "branch name is required")
		return uuid.Nil, "", false
	}
	return streamID, name, true
}

// validBranchName allows ASCII alphanumerics plus '-', '_', '/' —
// matches the Rust `c.is_ascii_alphanumeric() || c == '-' || c == '_' || c == '/'`.
func validBranchName(name string) bool {
	if name == "" {
		return false
	}
	for i := 0; i < len(name); i++ {
		c := name[i]
		switch {
		case c >= '0' && c <= '9':
		case c >= 'a' && c <= 'z':
		case c >= 'A' && c <= 'Z':
		case c == '-' || c == '_' || c == '/':
		default:
			return false
		}
	}
	return true
}

func branchActor(c *authmw.Claims) string {
	if c.Email != "" {
		return c.Email
	}
	return c.Sub.String()
}

func emitBranchAudit(c *authmw.Claims, action string, resourceID uuid.UUID, extra ...slog.Attr) {
	attrs := []any{
		slog.String("action", action),
		slog.String("actor_sub", c.Sub.String()),
		slog.String("actor_email", c.Email),
		slog.String("resource", "streaming_stream_branch"),
		slog.String("resource_id", resourceID.String()),
	}
	for _, a := range extra {
		attrs = append(attrs, a)
	}
	slog.Info("audit "+action, attrs...)
}

// Sentinel kept here so cold bridges that need to flag a "branch was
// already archived" race can reach a typed error if needed in future.
var ErrBranchAlreadyArchived = errors.New("stream branch already archived")
